text/2523-cfg-path-version.md

```diff
@@ -0,0 +1,956 @@
+Today, crate authors often face a dilemma: - *"Shall I provide more features
+and implementations for later versions of Rust, or should I stay compatible
+with more versions of the compiler"*.
+
+[cargo_version_selection]: http://aturon.github.io/2018/07/25/cargo-version-selection/
+
+While [much thought][cargo_version_selection] has been given to how we can
+reduce "dependency hell" by enhancing cargo for:
+
++ the **control** users have over their dependencies.
++ the **compatibility** of crates with each other.
++ reducing the **maintainability** burden of having to make sure that
+  versions work with each other.
+
+[RFC 2483]: https://github.com/rust-lang/rfcs/pull/2483
+
+...not much focus has been given to how conditional compilation can be improved
+to extend how many versions back a crate supports. This becomes critically
+important if and when we gain LTS channels as proposed by [RFC 2483].
+
+[version_check]: https://crates.io/crates/version_check
+
+The current support for such conditional compilation is lacking.
+While [it is possible][version_check] to check if you are above a certain
+compiler version, such facilities are not particularly ergonomic at the moment.
+In particular, they require the setting up of a `build.rs` file and
+declaring up-front which versions you are interested in knowing about.
+These tools are also unable to check, without performing canary builds
+of simple programs with `use ::std::some::path;`, if a certain path exists
+and instead force you to know which version they were introduced in.
+
+*We can do better.* In this RFC we aim to rectify this by giving library
+authors the tools they need in the language itself. With the features
+proposed in the [summary] we aim to make retaining *compatibility* and
+supporting more compiler versions *pain-free* and to give authors a lot
+of *control* over what is supported and what is not.
+
+[rust-lang-nursery/error-chain#101]: https://github.com/rust-lang-nursery/error-chain/issues/101
+
+Another use case this RFC supports is to work around compiler bugs by
+checking if we are on a particular version. An example where this occurred
+is documented in [rust-lang-nursery/error-chain#101].
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[cfg(accessible($path))]`
+
+Consider for a moment that we would like to use the `Iterator::flatten` method
+of the standard library if it exists (because it has become soon in a certain 
+Rust version), but otherwise fall back to `Itertools::flatten`.
+We can do that with the following snippet:
+
+```rust
+#[cfg(accessible(::std::iter::Flatten))]
+fn make_iter(limit: u8) -> impl Iterator<Item = u8> {
+    (0..limit).map(move |x| (x..limit)).flatten()
+}
+
+#[cfg(not(accessible(::std::iter::Flatten)))]
+fn make_iter(limit: u8) -> impl Iterator<Item = u8> {
+    use itertools::Itertools;
+    (0..limit).map(move |x| (x..limit)).flatten()
+}
```

Comment:
Should we show an alternative simpler code here?
```rust
fn make_iter(limit: u8) -> impl Iterator<Item = u8> {
    #[cfg(not(accessible(::std::iter::Flatten)))]
    use itertools::Itertools;
    (0..limit).map(move |x| (x..limit)).flatten()
}
```

Reply:
Sure, why not. Want to file a patch?

Reply:
I have created a pull request in https://github.com/rust-lang/rfcs/pull/2779.